distributed/mappings.py

```diff
@@ -128,36 +193,143 @@ def init_ddp_model_and_reduction_hooks(model,
+        # no model parallel, just use DDP with 
+        # the full world size
+        ddp_group = None
+    elif comm.get_size("cp") == 1:
+        # only cp requires additional allreduce
+        # if no cp, use DDP 
+        ddp_group = comm.get_group("dp")
+    else:
+        broadcast_buffers = False
+        ddp_group = comm.get_group("dp")
+        need_hooks = True # need a grad hook for additional reduce
+
+    model = DistributedDataParallel(model,
+                                    device_ids = device_ids,
+                                    output_device = output_device,
+                                    bucket_cap_mb = bucket_cap_mb,
+                                    broadcast_buffers = broadcast_buffers,
+                                    find_unused_parameters = find_unused_parameters,
+                                    gradient_as_bucket_view = gradient_as_bucket_view,
+                                    static_graph = static_graph,
+                                    process_group = ddp_group)
+    if not need_hooks:
+        return model
+
+    # define comm hook because some params need additional allreduce
+    def reduction_comm_hook(state: object, bucket: dist.GradBucket) -> torch.futures.Future[torch.Tensor]:
+        group_grads_map = {}
+        params = bucket.parameters()
+
+        for param in params:
+            if param.grad is not None:
+                grad = param.grad.data
+
+                if param.mark_for_reduction:
+                    # create combined group name (eg: 'dp-cp', 'dp-tp', etc)
+                    combined_group = "dp-" + "-".join(sorted(param.mark_for_reduction))
+                else:
+                    # default to 'dp' if no extra reduction is needed
+                    combined_group = "dp"
+
+                if combined_group not in group_grads_map:
+                    group_grads_map[combined_group] = []
+                group_grads_map[combined_group].append(grad)
+
+        allreduce_futures = []
+        dp_size = comm.get_size("dp")  # data parallel comm size
+
+        # perform asnc reduces on dp or combined dp-.. groups
+        for combined_group, grads in group_grads_map.items():
+            if grads:
+                coalesced = _flatten_dense_tensors(grads)
+                # divide by dp size since we need to average the grads
+                coalesced.div_(dp_size)
+                # Check if coalesced contains NaNs or Infs after division
+                if torch.isnan(coalesced).any() or torch.isinf(coalesced).any():
+                    raise ValueError(f"NaN or Inf detected in coalesced gradients for group: {combined_group}")
+
+                group = comm.get_group(combined_group)
+                # all-reduce in the appropriate combined group
+                fut_group = dist.all_reduce(coalesced, op=dist.ReduceOp.SUM, group=group, async_op=True).get_future()
+                # copy the results to the gradients (update the grads basically)
```

Comment:
This is the issue I was mentioning on the last call: the future is marked done when the reduction is enqueued, which in turn means that the grad buffer copy can fire off too early, leading to data corruption. Therefore, as in my original routine above, I recommend marking this one blocking (so not async) and then follow it up by a blocking copy. The performance impact is not very big I think.

The whole future business is not very well documented unfortunately and it took us quite a while to debug this logic. In my above logic, I was also using composite groups implicitly. For example if I have an image h,w decomposed, then I also have a "spatial" group which is the union of h and w. And in case weights are shared across h and w and need to be reduced, I am just using the parent group. So basically I decompose the communicator into a graph, with the most fine grained comms in the leaves.

Reply:
Ok got it. I spent some time in trying to understand the docs and thought the `then()` logic (https://pytorch.org/docs/stable/futures.html#torch.futures.Future.then) would ensure that the grad buffer copy would start only after the reduction is complete. So, the async reduction is enqueued, the future returns and the _sync_gradients is chained, but any operation on the gpu would start only after the prev op (the reduction) finishes. I could be very wrong.. (and think its painful to debug) and happy to revert to the blocking logic to be safe.

Reply:
Hmm, this might work. Have you tested this at larger scales and does it run?

Reply:
Please have a look here though: https://pytorch.org/docs/stable/_modules/torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.html#PowerSGDState

In this more complicated hook, they do not return the future from the async AR but also wait for it to complete. I think something like that needs to be done, but I am really not sure. That logic is very confusing.